sweep/walletsweep.go

```diff
@@ -260,6 +263,19 @@ func CraftSweepAllTx(feeRate, maxFeeRate chainfee.SatPerKWeight,
 
 		log.Trace("[WithCoinSelectLock] finished fetching UTXOs")
 
+		// Use select utxos, if provided.
+		if len(selectUtxos) > 0 {
+			if fn.HasDuplicates(selectUtxos) {
+				return errors.New("selected UTXOs contain " +
+					"duplicate values")
+			}
+
+			utxos, err = fetchUtxosFromOutpoints(utxos, selectUtxos)
```

Comment:
By the same argument, there's no need to check duplicates because an error will be returned?

Reply:
not really `fetchutxosFromOutpoints` does not check for duplicates. What I am trying to say is that if any select outpoint is unknown to the wallet, an error would be returned as the outpoint would not be found as seen here:
https://github.com/lightningnetwork/lnd/blob/ede12ddf588a14c865df00264be011954b563ddc/sweep/walletsweep.go#L422-L424

Reply:
It's simpler to show it in the test, here's what I did to hack an error case,
```diff
diff --git a/itest/lnd_misc_test.go b/itest/lnd_misc_test.go
index 4468a031e..45a261b5a 100644
--- a/itest/lnd_misc_test.go
+++ b/itest/lnd_misc_test.go
@@ -1536,6 +1536,8 @@ func testSendCoinSelectUtxo(ht *lntest.HarnessTest) {
 				)
 			}
 
+			selectedOutpoints[0].OutputIndex = 10
+
 			sendCoinReq := &lnrpc.SendCoinsRequest{
 				Addr:       addr.String(),
 				Outpoints:  selectedOutpoints,
```

And I got the error,
```
rpc error: code = Unknown desc = selected outpoint not eligible for spending: b4d01c803a271c2541401a2c1b38fc0e3c93c9d01267a519d5cc628a92b90169:10
```

I actually think the error returned is good enough. Got curious and went ahead to check what error would be returned when using duplicates,
```
rpc error: code = Unknown desc = bad txns inputs duplicate
```
Seems indicative enough.

This goes back to my original comment, that we only need one place to do the internal validation. However, it's also nice to validate user inputs on the RPC level, as it can save the computation effort down the road. So we should do it once in `SendCoins`, but not again in `CraftSweepAllTx`, again in `sendCoinsOnChain`, and again in `btcwallet`.

Reply:
I do not check for this in sendcoinsOnChain and btcwallet.go file though.  Unless by `btcwallet`, you mean this pull request here: https://github.com/btcsuite/btcwallet/pull/928/ Also you suggested that I used set in the `CraftSweepAllTx` and that is what I am implementing so there would be no need for the check. So we would only have the check in the `SendCoins` rpc function.

I would update this pull request: https://github.com/btcsuite/btcwallet/pull/928/commits to use set as well. If we do not want to use set, I still suggest that we carryout the check in these exported functions that could be used by anyone so that these functions do not return invalid transactions that would ultimately be rejected by the mempool.